streams/src/main/java/org/apache/kafka/streams/state/internals/LogicalKeyValueSegment.java

```diff
@@ -170,7 +173,30 @@ public boolean isOpen() {
 
     @Override
     public synchronized byte[] get(final Bytes key) {
-        return physicalStore.get(prefixKeyFormatter.addPrefix(key));
+        return get(key, Optional.empty());
+    }
+
+    public synchronized byte[] get(final Bytes key, final Snapshot snapshot) {
+        return get(key, Optional.of(snapshot));
+    }
+
+    private synchronized byte[] get(final Bytes key, final Optional<Snapshot> snapshot) {
+        if (snapshot.isPresent()) {
+            try (ReadOptions readOptions = new ReadOptions()) {
+                readOptions.setSnapshot(snapshot.get());
+                return physicalStore.get(prefixKeyFormatter.addPrefix(key), readOptions);
+            }
+        } else {
+            return physicalStore.get(prefixKeyFormatter.addPrefix(key));
+        }
+    }
+
+    public Snapshot getSnapshot() {
+        return physicalStore.db.getSnapshot();
```

Comment:
I think we should push this into `RocksDBStore` class, ie, call `physicalStore.getSnapshot()` here, and extend `RocksDBStore` to track all open snapshots (similar to how we track open iterators and release all open snapshot if the store is closed).

Reply:
> I think we should push this into `RocksDBStore` class, ie, call `physicalStore.getSnapshot()` here, and extend `RocksDBStore` to track all open snapshots (similar to how we track open iterators and release all open snapshot if the store is closed).

Since "Snapshots do not persist across DB restarts ([ref](https://github.com/facebook/rocksdb/wiki/Snapshot))", I think we do not need track closing them when closing the store.